### ISSUE TITLE
backup time on 0.9.x

### ISSUE THREAD

**Issue by mpetrazzo** [restic/restic#1840]
## Output of `restic version`
0.9.0

## How did you run restic exactly?
On windows server, I backup a big dataset (> 1ml small files, 250GB) and I find that that restic 0.9 take a lot of time to analyze the dataset though there isn't files modified or added.
On the same dataset, restic 0.8 takes an order of magnitude less: 0.9 3h, 0.8 2 minutes!

[See forum for details](https://forum.restic.net/t/optimize-backup-process-steps/721)

## What backend/server/service did you use to store the repository?
minio on LAN server

## Expected behavior
Restic need to run with the same performances (or better!) than the previous version 

## Steps to reproduce the behavior
Do a backup on a big dataset more than one time (the first time it's normal that take a lot of time since restic need to backup all the data).
The 2th, 3th and so on time that you run restic, it have to take not many time as the 1th. In my tests, it took more or less the sme time

P.s. I suggest to backup a big dataset since a small one isn't significant and may take a small amount of time for all the backups (with or without modification) and you could not see the differences in the execution time.

## Do you have any idea what may have caused this?
Maybe restic re-read all the data even if the file wasn't begin modified

## Do you have an idea how to solve the issue?
I tried to debug the process on a linux machine, but the behavior differs from the windows one and I didn't find what I was looking for.
With strace -p and a simple dataset, I backup two file for the 1th time, I modified only one and re-run the backup.
Restic seem to read completely only the modified file (that is correct), not explaining the problem as I was expecting (I thought that restic read all the data into the 2th run) 
I tried this with 0.8 and 0.9 and both do the same.

Attached you can find the tests (_1 it the first restic run, _2 the second)

[restic_8_1.txt](https://github.com/restic/restic/files/2087951/restic_8_1.txt)
[restic_8_2.txt](https://github.com/restic/restic/files/2087952/restic_8_2.txt)
[restic_9_1.txt](https://github.com/restic/restic/files/2087953/restic_9_1.txt)
[restic_9_2.txt](https://github.com/restic/restic/files/2087954/restic_9_2.txt)

This are the files / strings to search inside the strace output.
Restic was called with:
`strace -r -f -o /tmp/$filename`

First run / before modify the files:
```
michele@srv:/tmp/restic/source$ ls                                                                                                                                                   
modified  unmodified                  
michele@srv:/tmp/restic/source$ cat *
text_modified
text_unmodified
```
After:
```
michele@srv:/tmp/restic/source$ echo new_modified > modified 
michele@srv:/tmp/restic/source$ cat *
new_modified
text_unmodified
```

## Did restic help you or made you happy in any way?
Yes! Restic rocks. We are now trying it for use in production and our objective is to replace an old "enterprise" backup software with restic!

**Comment by LexxGit** [comment]
hi mpetrazzo and Alexander!

I had the exact same behaviour under Win7 64bit pro dataset: 520.000 files 18.000 directories 405GB

A second run with v0.8 took only 3.5 minutes while v0.9 took nearly 3 hours. SSD->SSD

Found out through resourcemonitor:
ESET Antivirus was heavily scanning many if not all Files of the dataset(source) and used  a lot of resources.

That would explain why you couldn't reproduce this behaviour with Linux (no Antivirus installed).

It took me some time to realize that the problem was the Antivirus-Engine, cause restic v0.8 didn't caused this behaviour.

Something must have marginally/heavily changed in the way restic touches/reads/checks even unmodified Files in the source folder between v0.8 and v0.9, causing the Antivirus engine to check the files. 

I turned off the ESET realtime protection of the sourcefolder.. 
bingo backuptime back to normal!

Hope this helps!
and btw restic ROCKS!

Alex

**Comment by mpetrazzo** [comment]
Hi,
here I'm with the windows tests.
I run the first and the second time restic, both version 0.8 and 0.9, with the same command line.
Here I report the 0.8 and 0.9 versions at the nth time (I haven't saved the output for the first and second run, but hope it helps) 


```
c:\Backup>"C:\Backup\restic_0.8.exe" -r s3:http://ipsave:9000/repo2-8 backup --cache-dir C:\Backup\cache\repo2-8  "D:\$mypath\repository2"
password is correct
using parent snapshot 9b18acf2
scan [D:\$mypath\repository2]
scanned 25 directories, 787299 files in 3:26
[1:16] 100.00%  299.381 GiB / 299.381 GiB  787324 / 787324 items  0 errors  ETA 0:00
duration: 1:16
snapshot 0927db18 saved
```
and 0.9:
```
c:\Backup>"C:\Backup\restic_0.9.exe" -r s3:http://ipsave:9000/repo2-9 backup --cache-dir C:\Backup\cache\repo2-9  "D:\$mypath\repository2"
repository f4e42162 opened successfully, password is correct

Files:           0 new,     0 changed, 787299 unmodified
Dirs:            0 new,     0 changed,     0 unmodified
Added:          0 GiB

processed 787299 files, 299.381 GiB in 3:30:57
snapshot 48d2049c saved
```

If you need, I can try to re-run 0.8 and 0.9 with a clean cache and repo

**Comment by LexxGit** [comment]
I forgot to mention, that i jumped directly from 0.8 to 0.9.

After finding the fact i tried 0.83 and had the same slow backups as with 0.9.

So the changes in the way restic "checks" the source and triggers ESET NOD Antivirus must have occured between 0.8 and 0.83.
hth
Alex

**Comment by mpetrazzo** [comment]
Interesting.
I didn't thought about it.
About restic tests version, I tried always the last version (0.8.3 and 0.9.0+0.9.1), starting before with the 0.9.x and after with the 0.8 (opposite to your 0.8 -> 0.9)

I'm going to try to disable AV on windows VM for the _"c:\backup\"_ folder and the "_restic*.exe_" executable and I'll let you know.

Thanks

**Comment by fd0** [comment]
The output looks fine, the `787299 unmodified` means that restic did the right thing and did not read all the files. I don't have any idea why it's slow, it should be fast. Very mysterious.

Please let us know if disabling AV helps...

**Comment by mpetrazzo** [comment]
Hi all,
just tried and... 2:05 minutes!
For test, tried another time with AV enabled: 3h

The strange thing it's that, with the same server and the same datafiles (and AV enabled), version 0.8 backtime is 3 or 4 minutes

Thanks

**Comment by fd0** [comment]
@mpetrazzo ok, so what I understand is that disabling the AV makes restic 0.9.0 fast again, while 0.8.3 is always fast. Is that correct?

In 0.9.0, we've completely replaced the archiver part, it now uses much less concurrent access to files. I can imagine that this looks very different from the outside, from the viewpoint of an AV program. Maybe with the new archiver we trigger some heuristics more easily.

However, I don't think there's anything we can do. I propose closing this issue. What do you think?

**Comment by dimejo** [comment]
> However, I don't think there's anything we can do. I propose closing this issue. What do you think?

Maybe add an entry to the FAQ about it?

**Comment by fd0** [comment]
Fine with me, is anyone up for a PR?

**Comment by dimejo** [comment]
> Fine with me, is anyone up for a PR?

I've added a PR but please feel free to expand on it as I think my english skills aren't good enough for official documentation.

**Comment by fd0** [comment]
Fixed in #1845, thanks @dimejo!
